tc/core/polyhedral/schedule_transforms.cc

```diff
@@ -797,5 +798,42 @@ void orderAfter(ScheduleTree* root, ScheduleTree* tree, isl::union_set filter) {
   parent->insertChild(childPos, std::move(seq));
 }
 
+/*
+ * Extract a mapping from the domain elements active at "tree"
```

Comment:
My comments are from legacy, so my apologies for raising this late.

First, I find it confusing to not see anything related to `activeDomainPoints`.
The name `extractMappingExpressions` feels more appropriate to me since you are really collecting the expressions for each mapping, irrespective of the sets of points.

Second, what happens in an imperfectly nested or full / partial tile world where you could generally have multiple mapping expressions to different subpieces of a given statement?
Doesn't this warrant some sanity checks and a comment?

Reply:
(Moving from main thread to scoped location in github UI)
> First, I find it confusing to not see anything related to `activeDomainPoints`.

Either name sounds reasonable to me.

> Second, what happens in an imperfectly nested or full / partial tile world where you could generally have multiple mapping expressions to different subpieces of a given statement?

The function will collect the mappings over the subpieces.

> Doesn't this warrant some sanity checks and a comment?
There is a comment:
 * where all branches in "tree"
 * are assumed to have been mapped to these identifiers.

and Alex added a sanity check in the caller.

skimo

Reply:
> The function will collect the mappings over the subpieces.

@skimo-openhub I'm sorry but I don't see it, here is what the function computes on `MatMulBias.RegistersAtRoot`:

```
E0610 11:21:34.416007 61796 schedule_transforms.cc:842] [K, M, N] -> block[{ S_0[O_s0_i, O_s0_j] -> [(O_s0_j - 32*floor((O_s0_j)/32))]; S_1[O_s1_i, O_s1_j, O_s1_k] -> [(O_s1_j - 32*floor((O_s1_j)/32))]; S_2[O_s2_i, O_s2_j] -> [(O_s2_j - 32*floor((O_s2_j)/32))] }, { S_0[O_s0_i, O_s0_j] -> [(O_s0_i - 8*floor((O_s0_i)/8))]; S_1[O_s1_i, O_s1_j, O_s1_k] -> [(O_s1_i - 8*floor((O_s1_i)/8))]; S_2[O_s2_i, O_s2_j] -> [(O_s2_i - 8*floor((O_s2_i)/8))] }]
```

Since no set/domain information is included (no active domain), there is an assumption that a statement may appear only in a single subtree below the current node (or that all mapping expressions for a statement must agree). Otherwise you would get multiple entries for the same statement and they would be impossible to disambiguate. The reusable functions in schedule_transforms should be generally applicable; here it seems they can only be applied under assumptions. So I think this needs a check for the above assumption and a comment. We may want to review other functions in this schedule_transforms.cc.

> Either name sounds reasonable to me

I've thought about this more I really think domain and active element should go (there is no domain or active elements in the computation); it's really just about recovering the expressions of the mappings.

Lastly it would be better to use the github web UI for review otherwise these comments get decoupled from their place of occurrence.

Reply:
On Sun, Jun 10, 2018 at 10:38:37AM -0700, Nicolas Vasilache wrote:
> nicolasvasilache commented on this pull request.
> > @@ -797,5 +798,42 @@ void orderAfter(ScheduleTree* root, ScheduleTree* tree, isl::unio
n_set filter) {
>    parent->insertChild(childPos, std::move(seq));
>  }
>
> +/*
> + * Extract a mapping from the domain elements active at "tree"
>
> > The function will collect the mappings over the subpieces.
>
> @skimo-openhub I'm sorry but I don't see it, here is what the function computes on `MatM
ulBias.RegistersAtRoot`:
>
> ```
> E0610 11:21:34.416007 61796 schedule_transforms.cc:842] [K, M, N] -> block[{ S_0[O_s0_i,
 O_s0_j] -> [(O_s0_j - 32*floor((O_s0_j)/32))]; S_1[O_s1_i, O_s1_j, O_s1_k] -> [(O_s1_j -
32*floor((O_s1_j)/32))]; S_2[O_s2_i, O_s2_j] -> [(O_s2_j - 32*floor((O_s2_j)/32))] }, { S_
0[O_s0_i, O_s0_j] -> [(O_s0_i - 8*floor((O_s0_i)/8))]; S_1[O_s1_i, O_s1_j, O_s1_k] -> [(O_
s1_i - 8*floor((O_s1_i)/8))]; S_2[O_s2_i, O_s2_j] -> [(O_s2_i - 8*floor((O_s2_i)/8))] }]
> ```

This in itself doesn't show much, but it seems you are right that
the mapping does not have enough information in general 
(in this case it does).

> Since no set/domain information is included (no active domain), there is an assumption t
hat a statement may appear only in a single subtree below the current node (or that all ma
pping expressions for a statement must agree). Otherwise you would get multiple entries fo
r the same statement and they would be impossible to disambiguate. The reusable functions
in schedule_transforms should be generally applicable; here it seems they can only be appl
ied under assumptions. So I think this needs a check for the above assumption and a commen
t. We may want to review other functions in this schedule_transforms.cc.

I was assuming that the mapping would contain information about the filters
because that is what I'm used to in PPCG, but it seems we do indeed 
not carry enough information in general.
I also misunderstood the check you wanted to add.
AFAIU now, you want to check that there is no point that gets mapped
more than once (rather than checking that each point gets mapped).

> > Either name sounds reasonable to me
>
> I've thought about this more I really think domain and active element should go

I don't understand.  Do you mean that these things should be removed?
Where should they be removed from?

> (there is no domain or active elements in the computation); it's really just about recov
ering the expressions of the mappings.
>
> Lastly it would be better to use the github web UI for review otherwise these comments g
et decoupled from their place of occurrence.

Ugh.  The github UI is horrible.  I'll just prepare an email and then paste it.

Reply:
I opened up #505; I have an implementation too, but it exposes some issue in isl
that I need to send out first.

Reply:
> AFAIU now, you want to check that there is no point that gets mapped more than once (rather than checking that each point gets mapped).

My concern with the information currently produced was that even totally disjoint sets of points getting mapped with different mappings would create ambiguities. So with the current information it's really: enforce <= 1 mapping filter per statement in the subtree. But what you write makes sense to

Reply:
I could not fix it because of the problem in isl.  I'm fine waiting for it to be fixed, or merging this and addressing in a separate PR